node/core/dispute-coordinator/src/real/mod.rs

```diff
@@ -499,10 +503,182 @@ async fn handle_new_activations(
+			RuntimeApiRequest::FetchOnChainVotes(tx),
+		))
+		.await;
+		match rx.await {
+			Ok(Ok(Some(val))) => val,
+			Ok(Ok(None)) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					"No on chain votes stored for relay chain leaf");
+				return Ok(())
+			},
+			Ok(Err(e)) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					error = ?e,
+					"Could not retrieve on chain votes due to an API error");
+				return Ok(())
+			},
+			Err(e) => {
+				tracing::trace!(
+					target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					error = ?e,
+					"Could not retrieve onchain votes due to oneshot cancellation");
+				return Ok(())
+			},
 		}
+	};
+
+	if backing_validators.is_empty() && disputes.is_empty() {
+		return Ok(())
 	}
 
+	// Obtain the session info, for sake of `ValidatorId`s
+	// either from the rolling session window or query from
+	// the on chain state directly
+	let session_info: SessionInfo =
+		if let Some(session_info) = state.rolling_session_window.session_info(session) {
+			session_info.clone()
+		} else {
+			let (tx, rx) = oneshot::channel();
+			ctx.send_message(RuntimeApiMessage::Request(
+				new_leaf,
+				RuntimeApiRequest::SessionInfo(session, tx),
+			))
+			.await;
+
+			match rx.await?? {
+				None => return Ok(()),
+				Some(session_info) => session_info,
+			}
+		};
+
+	// Scraped on-chain backing votes for the candidates with
+	// the new active leaf as if we received them via gossip.
+	for (candidate_receipt, backers) in backing_validators {
+		let candidate_hash = candidate_receipt.hash();
+		let statements = backers.into_iter().filter_map(|(validator_idx, validator_signature)| {
+			let validator_public: ValidatorId =
+				session_info.validators.get(validator_idx.0 as usize)?.clone();
+			let signed_dispute_statement = SignedDisputeStatement::new_checked(
+				DisputeStatement::Valid(ValidDisputeStatementKind::BackingValid(new_leaf)),
+				candidate_hash,
+				session,
+				validator_public,
+				validator_signature,
+			)
+			.ok()?;
+			Some((signed_dispute_statement, validator_idx))
+		});
+		let import_result = handle_import_statements(
+			ctx,
+			overlay_db,
+			state,
+			candidate_hash,
+			candidate_receipt,
+			session,
+			statements,
+			now,
+			metrics,
+		)
+		.await?;
+		match import_result {
+			ImportStatementsResult::ValidImport => tracing::trace!(target: LOG_TARGET,
+				relay_parent = ?new_leaf,
+				"Imported backing vote from on-chain"),
+			ImportStatementsResult::InvalidImport => tracing::warn!(target: LOG_TARGET,
+					relay_parent = ?new_leaf,
+					"Attempted import of on-chain backing votes failed"),
+		}
+	}
+
+	if disputes.is_empty() {
+		return Ok(())
+	}
+
+	// Import concluded disputes from on-chain, this already went through a vote so it's assumed
+	// as verified. This will only be stored, gossiping it is not necessary.
+
+	// First try to obtain all the backings which ultimately contain the candidate
```

Comment:
This seems wrong. It seems that the goal is to get the full candidate receipt, but I don't think that the candidate backing subsystem is the way to do it - the candidate has likely been backed at least a few blocks ago and the relevant job (which would be `candidate_receipt.relay_parent`, not `new_leaf`) has been pruned a while ago.

Reply:
That is unfortunately correct. Added a draft on how this might work, pending a sanity check with a fresh pair of eyes.